GVFS/GVFS.Virtualization/Projection/GitIndexProjection.SortedFolderEntries.cs

```diff
@@ -88,27 +88,57 @@ public void Clear()
                 this.sortedEntries.Clear();
             }
 
-            public FolderData AddFolder(LazyUTF8String name)
+            public FileData AddFile(LazyUTF8String name, byte[] shaBytes)
             {
                 int insertionIndex = this.GetInsertionIndex(name);
-                return this.InsertFolder(name, insertionIndex);
+                return this.InsertFile(name, shaBytes, insertionIndex);
             }
 
-            public FileData AddFile(LazyUTF8String name, byte[] shaBytes, bool isIncluded)
+            public FolderData GetOrAddFolder(
+                LazyUTF8String[] pathParts,
+                int partIndex,
+                bool parentIsIncluded,
+                SparseFolderData rootSparseFolderData)
             {
-                int insertionIndex = this.GetInsertionIndex(name);
-                return this.InsertFile(name, shaBytes, isIncluded, insertionIndex);
-            }
-
-            public FolderData GetOrAddFolder(LazyUTF8String name)
-            {
-                int index = this.GetSortedEntriesIndexOfName(name);
+                int index = this.GetSortedEntriesIndexOfName(pathParts[partIndex]);
                 if (index >= 0)
                 {
                     return (FolderData)this.sortedEntries[index];
                 }
 
-                return this.InsertFolder(name, ~index);
+                bool isIncluded = true;
+                if (rootSparseFolderData.Children.Count > 0)
+                {
+                    if (parentIsIncluded)
+                    {
+                        // Need to check if this child folder should be included
+                        SparseFolderData folderData = rootSparseFolderData;
+                        for (int i = 0; i <= partIndex; i++)
+                        {
+                            if (folderData.IsRecursive)
+                            {
+                                break;
+                            }
+
+                            string childFolderName = pathParts[i].GetString();
+                            if (!folderData.Children.ContainsKey(childFolderName))
+                            {
+                                isIncluded = false;
+                                break;
+                            }
+                            else
+                            {
+                                folderData = folderData.Children[childFolderName];
+                            }
+                        }
+                    }
+                    else
+                    {
+                        isIncluded = false;
+                    }
+                }
+
+                return this.InsertFolder(pathParts[partIndex], ~index, isIncluded: isIncluded);

```

Comment:
If the entry is not found the index is set the in the complement of the value where it should be inserted.  This is so there only needs to be one call to look something up instead of looking it up to discover it isn't found and looking up the index where it should be inserted when it isn't found.

Reply:
Thanks that makes sense